### ISSUE TITLE
Return same buffer if same object requested

### ISSUE THREAD

**Issue by danyill** [isomorphic-git/isomorphic-git#1105]
In Antora where `isomorphic-git` is delightfully used (thanks for this project and for supporting OSS), a large number of blobs are read from repositories as Antora analyzes the documentation requirements and determines which files should be retrieved from the repository. The nature of documentation is that not all of the content will have changed between different versions (which are typically different branches/tags).

Antora might be aggregating content from several branches or tags and in each branch (or tag), some of the files may still have the same git hash of the object in the index (what we would see if we did `git ls-files -s $file` 

Antora stores these all in memory (in a VFS) by doing something like:

```js
function entryToFile (entry) {
  return git.readBlob(entry).then(({ blob: contents }) => {
    const stat = new fs.Stats()
    stat.mode = entry.mode
    stat.mtime = undefined
    stat.size = contents.length
    return new File({ path: entry.path, contents, stat })
  })
}
```

where  `readBlob` is the isomorphic-git interface. 

Because some of these files are duplicates, Antora could be recording the git hash of the file in the index and deduplicating with some additional intelligence (effectively, referring to the same `Buffer` object). This would give the benefit of not using `n` times as much memory as there are tags/branches for an unchanging object.

At the moment Antora simply asks isomorphic-git for the same object again (source above can be found [here](https://gitlab.com/antora/antora/-/blob/master/packages/content-aggregator/lib/aggregate-content.js#L484-492)).

However what might be even cooler, would be that if isomorphic-git had already provided a Buffer with the same object hash previously it would return that same Buffer. This would give automatic deduplication of objects (and would make Antora more efficient, especially for binary objects).

**Comment by mojavelinux** [comment]
I agree, this was have a drastic impact on the amount of memory that Antora uses for sites that have a large number of files which share the same oid, a very likely scenario.

**Comment by djencks** [comment]
Note that this is using isomorphic-git 0.78.5, thus the reference to Buffer.  Since IIUC uint8Arrays are mutable I would think a flag to turn this on would be necessary to avoid users interfering with each other.

**Comment by mojavelinux** [comment]
While I love this idea, the more I think about it, the more I believe this is the responsibility of application, not isomorphic-git. If the application keeps track of the oid of file objects it has read, then it should be able to defer the call to readBlob and just use the value from its own index.

The only benefit of doing something like this in isomorphic-git would be that it already synchronizes access to the index cache. But that's something the application could emulate (though any advice would be welcome).

**Comment by wmhilton** [comment]
The main issue with trying to do caching at the library (isomorphic-git) level is determining how much. Because in memory-constrained environments, remembering all those buffers could eat up a lot of memory.

Right now, isomorphic-git doesn't cache any of the blobs at all. I think it should be easy enough to memoize `readBlob` at the application (antora) level. If it's requesting blobs by `oid` then memoize based on that. If it's requesting blobs by `filepath` then you can do `readTree` on the parent filepath and grab the `oid`s of the blob from there.

Orrrrr if all you care about is total memory usage and don't need a speedup by skipping the `readBlob` call... just do the `readBlob` calls as normal, and _then_ dedupe by `oid` and rely on the garbage collector to dispose the unused duplicate buffers:

```diff
+ const map = new Map()
  function entryToFile (entry) {
-   return git.readBlob(entry).then(({ blob: contents }) => {
+   return git.readBlob(entry).then(({ blob: contents, oid }) => {
+     contents = map.get(oid) || contents
+     map.set(oid, contents)
      const stat = new fs.Stats()
      stat.mode = entry.mode
      stat.mtime = undefined
      stat.size = contents.length
      return new File({ path: entry.path, contents, stat })
    })
  }
```
